### PR TITLE
Support conditional types

### PR DESCRIPTION
**Pull request by rvanvelzen**
Part of implementation for phpstan/phpstan#3853

### CODE REVIEW

File: src/Ast/Type/ConditionalTypeNode.php [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+	use NodeAttributes;
+
+	/** @var TypeNode */
+	public $subjectType;
+
+	/** @var TypeNode */
+	public $targetType;
+
+	/** @var TypeNode */
+	public $trueType;
```
Comment (ondrejmirtes) [inline src/Ast/Type/ConditionalTypeNode.php]: I'd rather have `if` type and `else` type, like here: https://github.com/nikic/PHP-Parser/blob/master/lib/PhpParser/Node/Expr/Ternary.php

File: src/Parser/TokenIterator.php [hunk 2]
```diff
@@ -59,7 +60,7 @@ public function currentTokenOffset(): int
 
 	public function isCurrentTokenValue(string $tokenValue): bool
 	{
-		return $this->tokens[$this->index][Lexer::VALUE_OFFSET] === $tokenValue;
+		return strcasecmp($this->tokens[$this->index][Lexer::VALUE_OFFSET], $tokenValue) === 0;
```
Comment (ondrejmirtes) [inline src/Parser/TokenIterator.php]: What is this change for?

File: src/Parser/TypeParser.php [hunk 3]
```diff
@@ -44,7 +46,9 @@ public function parse(TokenIterator $tokens): Ast\Type\TypeNode
 	private function parseAtomic(TokenIterator $tokens): Ast\Type\TypeNode
 	{
 		if ($tokens->tryConsumeTokenType(Lexer::TOKEN_OPEN_PARENTHESES)) {
+			$tokens->tryConsumeTokenType(Lexer::TOKEN_PHPDOC_EOL);
```
Comment (ondrejmirtes) [inline src/Parser/TypeParser.php]: This surely has some repercussions for some other parsed types. Can you add some tests demonstrating that?

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
Also please check this article: https://psalm.dev/articles/conditional-love

From what I understand, you implemented this syntax in this PR:

```php
   * @psalm-return (
   *     T is int
   *     ? static
   *     : array<static>
   * ) 
```

I'd also like to support this:

```php
   * @psalm-return (
   *     $id is int
   *     ? static
   *     : array<static>
   * ) 
```

Since `$id` is not a type, feel free to do it as a completely different AST node.

Thanks! This looks promising.

File: src/Parser/TokenIterator.php (hunk 2 above)
Comment (rvanvelzen) [reply src/Parser/TokenIterator.php]: This supports e.g. `T IS NOT NULL`. If you prefer only to support lowercase (i.e. `T is not null`) I'll revert this

Comment (ondrejmirtes) [reply src/Parser/TokenIterator.php]: Yes, please revert, I'm fine with supporting only a single variant :)

File: src/Parser/TypeParser.php (hunk 3 above)
Comment (rvanvelzen) [reply src/Parser/TypeParser.php]: Sure. I could also provide a separate PR for this change?

**Comment by rvanvelzen** [comment]
> I'd also like to support this:
> 
> ```
>    * @psalm-return (
>    *     $id is int
>    *     ? static
>    *     : array<static>
>    * ) 
> ```


I was planning to implement that syntax separately in a follow-up PR, but I'm fine with including it here as well.

> Since `$id` is not a type, feel free to do it as a completely different AST node.

Do you mean a separate `ConditionalTypeForVariable` node with a plain `$subjectName` property, or have `$subject` in `ConditionalType` be a union of `TypeNode|ParameterReferenceNode`? (All names up for debate).

**Comment by ondrejmirtes** [comment]
Separate `ConditionalTypeForParameter` with a plain string `$parameterName` property :) Thank you very much.

**Comment by rvanvelzen** [comment]
I believe I addresses everything. Please let me know if more changes are necessary

**Review by ondrejmirtes — APPROVED** [review]

**Comment by ondrejmirtes** [comment]
Thank you!

File: doc/grammars/type.abnf [hunk 4]
```diff
@@ -6,19 +6,26 @@ Type
...
+	/ Nullable
+
 Union
 	= 1*(TokenUnion Atomic)
 
 Intersection
 	= 1*(TokenIntersection Atomic)
 
+Conditional
+    = 1*ByteHorizontalWs TokenIs [TokenNot] Atomic TokenNullable Atomic TokenColon Atomic
```
Comment (staabm) [inline doc/grammars/type.abnf:20]: not sure how/wheter these `abnf` file ares sensitive to whitespace, but the newly introduced lines use spaces, not tabs

**Pull request merged**